lib/Service/AbstractService.php

```diff
@@ -40,85 +32,21 @@ public function getClient()
-    {
-        return $this->request('get', $this->baseNestedPath($parentId, $nestedPath), $params, $opts);
-    }
-
-    protected function createObject($params, $opts)
-    {
-        return $this->request('post', $this->basePath(), $params, $opts);
-    }
-
-    protected function createNestedObject($nestedPath, $parentId, $params, $opts)
-    {
-        return $this->request('post', $this->baseNestedPath($parentId, $nestedPath), $params, $opts);
-    }
-
-    protected function deleteObject($id, $params, $opts)
-    {
-        return $this->request('delete', $this->instancePath($id), $params, $opts);
-    }
-
-    protected function deleteNestedObject($nestedPath, $parentId, $id, $params, $opts)
-    {
-        return $this->request('delete', $this->instanceNestedPath($parentId, $nestedPath, $id), $params, $opts);
-    }
-
-    protected function retrieveObject($id, $params, $opts)
-    {
-        return $this->request('get', $this->instancePath($id), $params, $opts);
-    }
-
-    protected function retrieveNestedObject($nestedPath, $parentId, $id, $params, $opts)
-    {
-        return $this->request('get', $this->instanceNestedPath($parentId, $nestedPath, $id), $params, $opts);
-    }
-
-    protected function updateObject($id, $params, $opts)
-    {
-        return $this->request('post', $this->instancePath($id), $params, $opts);
-    }
-
-    protected function updateNestedObject($nestedPath, $parentId, $id, $params, $opts)
-    {
-        return $this->request('post', $this->instanceNestedPath($parentId, $nestedPath, $id), $params, $opts);
-    }
-
     protected function request($method, $path, $params, $opts)
     {
         return $this->getClient()->request($method, $path, $params, $opts);
     }
 
-    protected function baseNestedPath($parentId, $nestedPath)
-    {
-        return $this->instancePath($parentId) . '/' . $nestedPath;
-    }
-
-    protected function instancePath($id)
-    {
-        if (null === $id || '' === \trim($id)) {
-            $msg = 'The resource ID cannot be null or whitespace.';
-
-            throw new \Stripe\Exception\InvalidArgumentException($msg);
-        }
-
-        return $this->basePath() . '/' . \urlencode($id);
-    }
-
-    protected function instanceNestedPath($parentId, $nestedPath, $id)
+    protected function buildPath($basePath, ...$ids)
     {
-        if (null === $id || '' === \trim($id)) {
-            $msg = 'The resource ID cannot be null or whitespace.';
+        foreach ($ids as $id) {
+            if (null === $id || '' === \trim($id)) {
+                $msg = 'The resource ID cannot be null or whitespace.';
 
-            throw new \Stripe\Exception\InvalidArgumentException($msg);
+                throw new \Stripe\Exception\InvalidArgumentException($msg);
+            }
         }
 
-        return $this->baseNestedPath($parentId, $nestedPath) . '/' . \urlencode($id);
+        return \sprintf($basePath, ...\array_map('\urlencode', $ids));
```

Comment:
It's used to form URLs containing IDs, e.g.:
```php
$this->buildPath('/v1/coupons/%s', $id)
```

It's safe since the format string (the base path) is always a hardcoded string, not a user-supplied parameter.